### DISCUSSION TITLE
manual limb darkening tables

### DISCUSSION THREAD

**Discussion by dmrowan** [phoebe-project/phoebe2#537] [Questions]
Hello, 

I am interested in using manual limb darkening coefficients for a system with component Teffs above the range of the default ck2004 and phoenix tables. 

I have interpolated the values I want from the [van Hamme (1993)](https://ui.adsabs.harvard.edu/abs/1993AJ....106.2096V/abstract) table for the square root law for the two components of my system. Call them x1, y1, and x2, y2. 

After setting up my bundle with the orbital and stellar parameters, these are the relevant steps I am taking with PHOEBE:

```
bundle.set_value_all('ld_mode', 'manual')
bundle.set_value_all('ld_mode_bol', 'manual')
bundle.set_value_all('atm', 'blackbody')

bundle.set_value('ld_func@primary', 'square_root')
bundle.set_value('ld_func@secondary', 'square_root')

bundle.set_value('ld_coeffs@primary', [x1, y1])
bundle.set_value('ld_coeffs@secondary', [x2, y2])

#for this system I also set the following:
bundle['gravb_bol@primary'] = 1.0
bundle['gravb_bol@secondary'] = 1.0
bundle['irrad_frac_refl_bol@primary'] = 1.0
bundle['irrad_frac_refl_bol@secondary'] = 1.0
```
I have four questions:
1. Should I have set ld_mode AND ld_mode_bol to manual? I am confused what the difference is. For my model I am using irradation, if that matters. 
2. Do I need to set the value of the ld_coeffs_bol? 
3. Is there a way to 'expose' the limb darkening law like we can see the expression for a component constraint? I want to be sure the coeffs I am using refer to the same coeffs in the PHOEBE implementation of the square root limb darkening law
4. If I want to do optimization for this system where the teff ratio is a free parameter, does it make more sense to do nelder mead optimization in "chunks"? For example, every 20 iterations should I change the limb darkening coeffs given the current Teff and logg of each component? Is there a better way to emulate the behavior of the ck2004/phoenix lookup mode for a custom table during optimization?

**Comment by Raindogjones** [comment]
Hi @dmrowan,
1. You *could* leave ld_mode_bol untouched, in which case it will use the values 'lookup' mode (that phoebe uses by default) to determine the limb-darkening coefficients to use for the bolometric case.  This is only used in deriving the irradiation which is treated as a bolometric effect (so needs to use bolometric rather than passband dependent limb-darkening).  The parameters used by 'lookup' are based on the model atmosphere you are using for the star (Castelli & Kurucz 2004 by default)
2. If you switch to 'manual' then yes, you need to do this.  You might also want to change ld_func_bol too.
3.  The definition of the limb-darkening laws/functions used by phoebe is in Section 5.2.3 of Prsa et al. (2016) https://ui.adsabs.harvard.edu/abs/2016ApJS..227...29P/abstract
4. This is probably a question for @kecnry or @gecheline, I think.

**Reply by kecnry** [reply]
Running in chunks is definitely the easiest option, and hopefully once you get near the solution you could just leave them fixed.  Your other option is to implement your own lookup within a [custom cost function](http://phoebe-project.org/docs/2.3/tutorials/emcee_custom_lnprob).  You would keep the actual returned cost function as-is, but first setting ld coefficients based on the temperature, etc, and then **re-running** the forward model so that it accounts for those changes.  This unfortunately will double the cost as there is no way to skip the initial forward model.

If a similar setup but allowing for passing a function that gets executed **pre-compute** would be useful, that is something we could look into as a feature request moving forward.  I don't think that would be all that difficult to get up and running if you'd be interested and willing to test for your use-case?  I would imagine something like `custom_precompute_callable` which would allow you to change parameter values that cannot be done with constraints *before* the model is computed at each iteration.

**Reply by dmrowan** [reply]
Thank you both!

I found that if I don't set ld_mode_bol to manual I get the following error:

> Could not lookup ld_coeffs for ld_coeffs_source_bol@primary@star@component.  Try changing ld_coeffs_source_bol to a table that covers a sufficient range of values or set ld_mode_bol to 'manual' and manually provide coefficients via ld_coeffs_bol

And changing to manual (as I did in my original post) fixes this. This out of bounds error is, as described in my reponse to @aprsa below, due to the teff and logg combination. So I think if I use the van Hamme table I need to interpolate BOTH the passband dependent and bolometric limb darkening coefficients for the teffs and loggs of each component. Does that seem right?

@kecnry -- I think I am following the pre-compute suggestion. Right now I have a function that interpolates over a table of limb darkening coeffs and returns the coeffs for a given teff and logg. Would a feature like custom_precompute_callable take that function to set the ld coeffs before each evaluation in the optimization? I would be happy to test this -- even for a "normal" case where the atmosphere parameters were in the range of the ck2004 table it would be nice to have the option to set ld coeffs manually that update with the new atmosphere params.

**Reply by kecnry** [reply]
@dmrowan - the `custom_precompute_callable` would probably just allow you to edit any parameters in the bundle, using whatever logic you'd like, between when the optimizer/sampler sets its proposed values and the internal call to `run_compute`.  Obviously this could cause some poor optimization behavior if abused, but I think for cases like this it could make sense.  

The other thing I might try is to expose the ability for custom (non-algebraic) constraints.  That would arguably be more "clean" but comes with a bunch of portability complications since constraints are supposed to be serializable to an ascii file.

I'll try to implement one or both of these in a separate branch in the nearish future (a bit limited on time right now, so I don't dare promise a time frame) and give you instructions for trying it on your end and we can go from there.  In the meantime, you can start doing the same in `custom_lnprobability_callable` if you want and any code you write for the lookup should be easily ported to either of the other two options once implemented.

As @aprsa mentioned - for your case, this will likely all be safely avoided once transitioning from atmospheres to blackbody is released.  But I *think* having the flexibility to write custom logic might still be useful going forward for other cases as well.

**Comment by aprsa** [comment]
@dmrowan, I'm a bit confused about using Van Hamme for out-of-bounds atmosphere parameters. I thought Walter's tables covered the exact same range as ck2004; what are Teff/logg/abun that you need the coefficients for?

I might also mention that we're actively working on smooth, seamless extrapolation of both intensities and limb darkening coefficients as you transition from model atmosphere regions to blackbody regions. That will be included with the next major phoebe release.

**Reply by dmrowan** [reply]
The atmosphere parameters are teff=[42000.], logg=[3.67230036], abun=[0.] and return an out of bounds error for ldatam=ck2004. 

When I saw this error I assumed it was the Teff -- but I looking more carefully at table 1 of ck2004 I see that the problem is in the logg. I jumped to the Van Hamme tables based on previous literature for systems similar to the one I'm trying to model -- [Gamen09](https://ui.adsabs.harvard.edu/abs/2009A%26A...506.1269G/abstract) and [Bonanos04](https://ui.adsabs.harvard.edu/abs/2004ApJ...611L..33B/abstract), for example. I am not particularly attached to either table -- if there is a way to just take the ld coeffs at the nearest logg (looks like 4.5) for ck2004 that is fine with me. This is what I did for the van Hamme table. Do you have any suggestions for what I should do?

> I might also mention that we're actively working on smooth, seamless extrapolation of both intensities and limb darkening coefficients as you transition from model atmosphere regions to blackbody regions. That will be included with the next major phoebe release.

That will be very helpful!
